src/lib/firewall.ts

```diff
@@ -0,0 +1,198 @@
+import * as _ from 'lodash';
+
+import * as config from '../config/index';
+import * as constants from './constants';
+import * as iptables from './iptables';
+import { log } from './supervisor-console';
+
+import * as dbFormat from '../device-state/db-format';
+
+export const initialised = (async () => {
+	await config.initialized;
+	await applyFirewall();
+
+	// apply firewall whenever relevant config changes occur...
+	config.on('change', async ({ firewallMode, localMode }) => {
+		if (firewallMode || localMode !== undefined) {
+			applyFirewall({ firewallMode, localMode });
+		}
+	});
+})();
+
+const BALENA_FIREWALL_CHAIN = 'BALENA-FIREWALL';
+const LOG_PREFIX = '\u{1F525}';
+
+const prepareChain: iptables.Rule[] = [
+	{
+		action: '-N',
+	},
+	{
+		action: '-F',
+	},
+];
+
+const standardServices: iptables.Rule[] = [
+	{
+		comment: 'SSH Server',
+		action: '-A',
+		proto: 'tcp',
+		matches: ['--dport 22222'],
+		target: 'ACCEPT',
+	},
+	{
+		comment: 'balenaEngine',
+		action: '-A',
+		proto: 'tcp',
+		matches: ['--dport 2375'],
+		target: 'ACCEPT',
+	},
+	{
+		comment: 'mDNS',
+		action: '-A',
+		matches: ['-m addrtype', '--dst-type MULTICAST'],
+		target: 'ACCEPT',
+	},
+	{
+		comment: 'ICMP',
+		action: '-A',
+		proto: 'icmp',
+		target: 'ACCEPT',
+	},
+];
+
+const standardPolicy: iptables.Rule[] = [
+	{
+		comment: 'Locally-sourced traffic',
+		action: '-I',
+		matches: ['-m addrtype', '--src-type LOCAL'],
+		target: 'ACCEPT',
+	},
+	{
```

Comment:
What's this rule for?

Reply:
Locally-sourced traffic -- traffic coming into the device from itself; stuff like DNS lookups that use 127.0.0.2